docs/source/usage.rst

```diff
@@ -422,9 +422,13 @@ The following example sets ``abc123`` as the default value when ``DB_PASSWORD``
 Decoding strings with interpolations
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-You can automatically convert a string to its corresponding type (e.g., bool, int, float, dict, list) using ``oc.decode``
-(which can even resolve interpolations).
-This resolver also accepts ``None`` as input, in which case it returns ``None``.
+Strings may be converted using ``oc.decode``:
+
+- Primitive values (e.g., ``"true"``, ``"1"``, ``"1e-3"``) are automatically converted to their corresponding type (bool, int, float)
+- Dictionaries and lists (e.g., ``"{a: b}"``, ``"[a, b, c]"``)  are returned as transient config nodes (DictConfig and ListConfig)
+- Interpolations (e.g., ``"${foo}"``) are automatically resolved
```

Comment:
Are interpolations resolved on access (lazily) of the interpolation field or eagerly on access of the oc.decode resolver?

Reply:
They are resolved when the `oc.decode` resolver is called. I'm not sure what would be the "lazy" alternative you mention though, since if `oc.decode` is called, it means we are accessing the field (so we have to resolve it). Can you clarify?

Reply:
Ok thanks, I think I see what you mean. It's going to depend on what `oc.decode` gets as input:
* If it is the string `${foo}` then it will resolve it and return the corresponding value (that may be of any type)
* If it is the string `\${foo}` then it will return the string `${foo}`, that may get resolved later dynamically as an interpolation depending on what is being done with it

So, the bottom line is that both are possible, depending on the escaping in the string that `oc.decode` is processing.

While messing with it I noticed that there are a couple of problems with the escaping in quoted strings => I'll open separate issues for these.

Reply:
> The second case is also possible without escaping if the interpolation is a return of a custom resolver (like oc.env), right?
> 
> ```python
> os.environ["FOO"] = "${a}"
> cfg = OmegaConf.create({ "a" : 10, "b": "${oc.decode:{'foo:${env:FOO}'}" })
> ```
> 
> I am expecting:
> 
> ```
> b = cfg.b
> # b == DictConfig({"foo" : "${a}"}, parent=cfg)
> ```

No your example won't work. First, here's a version that runs (yours has some syntax issues) but doesn't give the outome you were hoping for:

```python
os.environ["FOO"] = "${a}"
cfg = OmegaConf.create({ "a" : 10, "b": "${oc.decode:'{foo:${oc.env:FOO}}'}" })
cfg.b  # {'foo': 10}
```
The reason is that the string the `decode()` function sees is `"{foo: ${a}}"` (see my first bullet point above).

What you want to achieve can be done with the following, which feeds `decode()` with the string `"{foo: '\${a}'}"` (my second bullet point above):
```python
os.environ["FOO"] = "'\\${a}'"
cfg = OmegaConf.create({ "a" : 10, "b": "${oc.decode:'{foo:${oc.env:FOO}}'}" })
cfg.b  # {'foo': '${a}'}
```